cmake/Config.cmake

```diff
@@ -115,6 +100,30 @@ elseif(MSVC)
         set(SFML_MSVC_VERSION 12)
     elseif(MSVC_VERSION EQUAL 1900)
         set(SFML_MSVC_VERSION 14)
+    elseif(MSVC_VERSION LESS_EQUAL 1919)
+        set(SFML_MSVC_VERSION 15)
+    elseif(MSVC_VERSION LESS_EQUAL 1929)
+        set(SFML_MSVC_VERSION 16)
+    endif()
```

Comment:
According to [Wikipedia](https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering), MSVC++ version 14.* spans from `_MSC_VER` values 1900 (Visual Studio 2015 14.0) to 1929 (Visual Studio 2019 Version 16.9). 

Microsoft hasn't disappointed in adding even more parallel versioning schemes to confuse the hell out of the users.
For example, the latest version listed is `14.29 | 1929 (Visual Studio 2019 Version 16.9)`, which contains *four* (!) different versions:
* MSVC++ compiler version 14.29
* `_MSC_VER` 1929
* Visual Studio 2019 (product name)
* Visual Studio Version 16.9 (product version)

So the number `SFML_MSVC_VERSION` in our code seems to match the fourth of these version numbers, or at least the integral part of it. Do we actually gain anything from this, semantically -- i.e. is there a case differentiation based on this specific versioning scheme, out of the four?

I see we use `SFML_MSVC_VERSION GREATER_EQUAL 14` further below, but is this a more reliable/future-proof way rather than checking ` MSVC_VERSION GREATER_EQUAL 1900`? Or do we use it because we already maintain the version mapping anyway (which would also be OK).

Also, maybe let's link the Wikipedia article and document what this number is referring to, took me a while to figure that out...

Reply:
I think it's easier to deal with 14/15/16 instead of even more magic numbers, so I think the mapping still has its merits.
Historically, this mapping was essential, because every VS version required new extlibs builds, which isn't really necessary anymore since VS 2017.
In theory we could remove the old mappings and just have one for pre/post VS 2017

Reply:
> I think it's easier to deal with 14/15/16 instead of even more magic numbers, so I think the mapping still has its merits.

My point was not to introduce yet another mapping, but work directly with `MSVC_VERSION` rather than `SFML_MSVC_VERSION`. But if most of the differences happen on the 14/15/16 boundaries, then the current approach is fine.


---

Taking a step back, this complexity is really insane. I mean, we're not only constantly doing manual case differentiations on compilers and versions, but we also need to deal with custom version mappings, warnings differing on each compiler, despite using the "meta build system" CMake with the promise to abstract from those very differences.

The C++ ecosystem has really become a mess.

Reply:
Yeah, I understood what you meant. It's not a huge benefit, but still makes it easier in our "own" CMake code to see 14/15/16 instead of random version numbers. Currently I only know of the two cases where it's really used, 1) for detecting if UCRT (universal CRT) should be used or the other one and 2) whether legacy_stdio_definitions.lib need to be linked.

----

Some stuff could potentially be simplified, by using more CMake specific functions. The initial modernization wasn't refactoring all parts of the scripts. Plus we try to make it as much compatible and easy to use as possible, so we allow certain configurations, that you won't necessarily find in other build setups. In the end, I still believe it's a lot easier to deal with than with like autotools or similar.

Reply:
Absolutely, I think CMake is still the lesser evil, and for SFML users it's quite reasonable. 

It's just... once you look at other programming languages, you notice how extremely over-complicated C++ dependency management and build configuration are, and the whole toolchain fragmentation certainly doesn't help. Out of curiosity, I checked how much configuration code we have, hopefully with the right command:
```sh
git ls-files | grep -E "CMakeLists\\.txt|cmake/(Config|Macros)\\.cmake" | xargs wc -l
```
So 2.2 kLOC for CMake. You could write a decent little game in that amount of code 🙂